src/engine/controls/clockcontrol.cpp

```diff
@@ -30,27 +33,94 @@ void ClockControl::trackLoaded(TrackPointer pNewTrack) {
-    m_pCOBeatActive->set(0.0);
+    m_pCOBeatActive->forceSet(0.0);
     m_pBeats = pBeats;
 }
 
 void ClockControl::process(const double dRate,
-                           const double currentSample,
-                           const int iBuffersize) {
+        const double currentSample,
+        const int iBuffersize) {
+    Q_UNUSED(dRate);
+    Q_UNUSED(currentSample);
     Q_UNUSED(iBuffersize);
-    double samplerate = m_pCOSampleRate->get();
+}
+
+void ClockControl::updateIndicators(const double dRate,
+        const double currentSample) {
+    /* This method sets the control beat_active is set to the following values:
+    * +1.0 --> Forward playing, set at the beat and set back to 0.0 at 20% of beat distance
+    * +0.5 --> Direction changed to reverse playing while forward playing indication was on
+    *  0.0 --> No beat indication
+    * -0.5 --> Direction changed to forward playing while reverse playing indication was on
+    * -1.0 --> Reverse playing, set at the beat and set back to 0.0 at -20% of beat distance
+    */
 
     // TODO(XXX) should this be customizable, or latency dependent?
-    const double blinkSeconds = 0.100;
+    const double kBlinkInterval = 0.20; // LED is on 20% of the beat period
 
-    // Multiply by two to get samples from frames. Interval is scaled linearly
-    // by the rate.
-    const double blinkIntervalSamples = 2.0 * samplerate * (1.0 * dRate) * blinkSeconds;
+    if ((currentSample == m_lastEvaluatedSample) ||
+            (dRate == 0.0)) {
+        return; // No position change (e.g. deck stopped) -> No indicator update needed
+    }
 
     mixxx::BeatsPointer pBeats = m_pBeats;
     if (pBeats) {
-        double closestBeat = pBeats->findClosestBeat(currentSample);
-        double distanceToClosestBeat = fabs(currentSample - closestBeat);
-        m_pCOBeatActive->set(distanceToClosestBeat < blinkIntervalSamples / 2.0);
+        if ((currentSample >= m_NextBeatSamples) ||
+                (currentSample <= m_PrevBeatSamples)) {
+            //qDebug() << "### findPrevNextBeats ### " << " currentSample: " << currentSample << " m_lastEvaluatedSample: " << m_lastEvaluatedSample << " m_PrevBeatSamples: " << m_PrevBeatSamples << " m_NextBeatSamples: " << m_NextBeatSamples;
+
+            pBeats->findPrevNextBeats(currentSample,
+                    &m_PrevBeatSamples,
+                    &m_NextBeatSamples,
+                    true); // Precise compare without tolerance needed
+
+            m_blinkIntervalSamples = (m_NextBeatSamples - m_PrevBeatSamples) * kBlinkInterval;
+        }
+        //qDebug() << "dRate:" << dRate << " m_lastPlayDirection:" << m_lastPlayDirection << " m_pCOBeatActive->get(): " << m_pCOBeatActive->get() << " currentSample: " << currentSample << " m_lastEvaluatedSample: " << m_lastEvaluatedSample << " m_PrevBeatSamples: " << m_PrevBeatSamples << " m_NextBeatSamples: " << m_NextBeatSamples << " m_blinkIntervalSamples: " << m_blinkIntervalSamples;
```

Comment:
Please break these long lines.

Reply:
> I am unsure if this is now just right. What is your use case for -1 -0.5 and 0.5?

The intention was to give the mapping developer all possibilities:
1)Search the beat by LED (therfore the LED must go of if direction changes)
2)Only show LED in normal playing direction.
3)LED On if position is close to the beat independent of direction

But maybe it's too much. If I remove the +- 0.5 states an use the value 0 instead, a simple
if (beat_active !== 0)
   set_led()
} else {
   reset_led()
}
would work for case 1.
Only case 3) would be impossible. The reason for case 3 is taht it's similar to the current behavio,

Reply:
I will remove this line later - it's just for my own debugging.